perception/probabilistic_occupancy_grid_map/src/pointcloud_based_occupancy_grid_map/occupancy_grid_map_fixed.cpp

```diff
@@ -83,40 +80,94 @@ void OccupancyGridMapFixedBlindSpot::updateWithPointCloud(
-  raw_pointcloud_angle_bins.resize(angle_bin_size);
-  for (PointCloud2ConstIterator<float> iter_x(scan_raw_pointcloud, "x"),
-       iter_y(scan_raw_pointcloud, "y"), iter_wx(map_raw_pointcloud, "x"),
-       iter_wy(map_raw_pointcloud, "y");
-       iter_x != iter_x.end(); ++iter_x, ++iter_y, ++iter_wx, ++iter_wy) {
-    const double angle = atan2(*iter_y, *iter_x);
-    const int angle_bin_index = (angle - min_angle) / angle_increment;
+  std::vector</*angle bin*/ std::vector<BinInfo>> obstacle_pointcloud_angle_bins(angle_bin_size);
+  std::vector</*angle bin*/ std::vector<BinInfo>> raw_pointcloud_angle_bins(angle_bin_size);
+  const int x_offset_raw = raw_pointcloud.fields[pcl::getFieldIndex(raw_pointcloud, "x")].offset;
+  const int y_offset_raw = raw_pointcloud.fields[pcl::getFieldIndex(raw_pointcloud, "y")].offset;
+  const int z_offset_raw = raw_pointcloud.fields[pcl::getFieldIndex(raw_pointcloud, "z")].offset;
+  const int x_offset_obstacle =
+    obstacle_pointcloud.fields[pcl::getFieldIndex(obstacle_pointcloud, "x")].offset;
+  const int y_offset_obstacle =
+    obstacle_pointcloud.fields[pcl::getFieldIndex(obstacle_pointcloud, "y")].offset;
+  const int z_offset_obstacle =
+    obstacle_pointcloud.fields[pcl::getFieldIndex(obstacle_pointcloud, "z")].offset;
+  const size_t raw_pointcloud_size = raw_pointcloud.width * raw_pointcloud.height;
+  const size_t obstacle_pointcloud_size = obstacle_pointcloud.width * obstacle_pointcloud.height;
+  const size_t raw_reserve_size = raw_pointcloud_size / angle_bin_size;
+  const size_t obstacle_reserve_size = obstacle_pointcloud_size / angle_bin_size;
+  // Reserve a certain amount of memory in advance for performance reasons
+  for (auto & raw_pointcloud_angle_bin : raw_pointcloud_angle_bins) {
+    raw_pointcloud_angle_bin.reserve(raw_reserve_size);
+  }
+  for (auto & obstacle_pointcloud_angle_bin : obstacle_pointcloud_angle_bins) {
+    obstacle_pointcloud_angle_bin.reserve(obstacle_reserve_size);
+  }
+  size_t global_offset = 0;
+  for (size_t i = 0; i < raw_pointcloud_size; ++i) {
+    Eigen::Vector4f pt(
+      *reinterpret_cast<const float *>(&raw_pointcloud.data[global_offset + x_offset_raw]),
+      *reinterpret_cast<const float *>(&raw_pointcloud.data[global_offset + y_offset_raw]),
+      *reinterpret_cast<const float *>(&raw_pointcloud.data[global_offset + z_offset_raw]), 1);
+    // Exclude invalid points
+    if (!std::isfinite(pt[0]) || !std::isfinite(pt[1]) || !std::isfinite(pt[2])) {
+      global_offset += raw_pointcloud.point_step;
+      continue;
+    }
+    // Apply height filter
+    if (pt[2] < min_height_ || max_height_ < pt[2]) {
+      global_offset += raw_pointcloud.point_step;
+      continue;
+    }
+    // Calculate transformed points
+    Eigen::Vector4f pt_map(matmap * pt);
+    Eigen::Vector4f pt_scan(matscan * pt_map);
+    const double angle = atan2(pt_scan[1], pt_scan[0]);
+    const int angle_bin_index = (angle - min_angle) * angle_increment_inv;
```

Comment:
Since these processes duplicate line 139--161, I recommended to unify them.

Reply:
I implemented this part as functions, but it slightly increased latency. So i'm thinking of the more efficient way.